File: LC_207.java
// 207. Course Schedule

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<Integer>[] graph = new ArrayList[numCourses];
        int[] inDegrees = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++)
            graph[i] = new ArrayList<>();
        
        for (int[] edge : prerequisites) {
            graph[edge[1]].add(edge[0]);
            inDegrees[edge[0]]++;
        }
        
        Queue<Integer> queue = new ArrayDeque<Integer>();
        
        for (int i = 0; i < numCourses; i++) {
            if (inDegrees[i] == 0)
                queue.offer(i);
        }
        
        int count = 0;
        
        while (!queue.isEmpty()) {
            int u = queue.poll();
            count++;
            
            for (Integer v : graph[u]) {
                if (--inDegrees[v] == 0) {
                    queue.offer(v);
                }
            }
        }
        
        return count == numCourses;
    }
}

File: LC_207.py
# 207. Course Schedule

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        in_degrees = [0] * numCourses
        
        for u, v in prerequisites:
            graph[v].append(u)
            in_degrees[u] += 1
            
        queue = collections.deque()
        
        for u in range(len(in_degrees)):
            if in_degrees[u] == 0:
                queue.append(u)
        
        count = 0
        while queue:
            u = queue.popleft()
            count += 1
            for v in graph[u]:
                in_degrees[v] -= 1
                if in_degrees[v] == 0:
                    queue.append(v)
        
        return count == numCourses
        